crates/core/src/node/inner/node_executor.rs

```diff
@@ -0,0 +1,449 @@
+        let result = if let Err(result) = reply.send(result) {
+            tracing::info!("failed to reply as receiver has been dropped");
+            result
+        } else {
+            return;
+        };
+        // Not much we can do with an error at this level so we just print it
+        if let Err(err) = result {
+            tracing::error!("failed to seal blocks: {:#?}", err);
+        }
+    }
+
+    async fn increase_time(
+        node_inner: Arc<RwLock<InMemoryNodeInner>>,
+        delta: u64,
+        reply: oneshot::Sender<()>,
+    ) {
+        node_inner.write().await.time_writer.increase_time(delta);
+        // Reply to sender if we can
+        if reply.send(()).is_err() {
+            tracing::info!("failed to reply as receiver has been dropped");
+        }
+    }
+
+    async fn enforce_next_timestamp(
+        node_inner: Arc<RwLock<InMemoryNodeInner>>,
+        timestamp: u64,
+        reply: oneshot::Sender<anyhow::Result<()>>,
+    ) {
+        let result = node_inner
+            .write()
+            .await
+            .time_writer
+            .enforce_next_timestamp(timestamp);
+        // Reply to sender if we can, otherwise hold result for further processing
+        let result = if let Err(result) = reply.send(result) {
+            tracing::info!("failed to reply as receiver has been dropped");
+            result
+        } else {
+            return;
+        };
+        // Not much we can do with an error at this level so we just print it
+        if let Err(err) = result {
+            tracing::error!("failed to enforce next timestamp: {:#?}", err);
+        }
+    }
+
+    async fn set_current_timestamp(
+        node_inner: Arc<RwLock<InMemoryNodeInner>>,
+        timestamp: u64,
+        reply: oneshot::Sender<i128>,
+    ) {
+        let result = node_inner
+            .write()
+            .await
+            .time_writer
+            .set_current_timestamp_unchecked(timestamp);
+        // Reply to sender if we can
+        if reply.send(result).is_err() {
+            tracing::info!("failed to reply as receiver has been dropped");
+        }
+    }
+
+    async fn set_timestamp_interval(node_inner: Arc<RwLock<InMemoryNodeInner>>, delta: u64) {
+        node_inner
+            .write()
+            .await
+            .time_writer
+            .set_block_timestamp_interval(Some(delta));
+    }
+
+    async fn remove_timestamp_interval(
+        node_inner: Arc<RwLock<InMemoryNodeInner>>,
+        reply: oneshot::Sender<bool>,
+    ) {
+        let result = node_inner
+            .write()
+            .await
+            .time_writer
+            .remove_block_timestamp_interval();
+        // Reply to sender if we can
+        if reply.send(result).is_err() {
+            tracing::info!("failed to reply as receiver has been dropped");
+        }
+    }
+}
+
+impl Future for NodeExecutor {
+    type Output = ();
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
```

Comment:
Maybe it's a stupid question, but why can't we just do a `run` method that would do something like this?
It feels a bit overengineered.

```
while let Some(command) = this.command_receiver.next().await {
  ...
}
```

Reply:
We can. I was slowly changing the original low-level implementation until I eventually arrived to what I have there right now and yeah what you are saying makes total sense. Reworking it, should be done by EOD.